Common/MemArenaPosix.cpp

```diff
@@ -107,12 +107,16 @@ void MemArena::ReleaseSpace() {
 
 void *MemArena::CreateView(s64 offset, size_t size, void *base)
 {
+#ifdef __EMSCRIPTEN__
+	void *retval = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_FILE | MAP_SHARED, fd, offset);
```

Comment:
Well, I'm not very interested in emscripten for PPSSPP one way or another, so I won't get in your way.  But is this safe?

The intent of this function is to create address space that matches PSP address space.  That is, when PSP code accesses these addresses, it refers to exactly the same memory:

0x08800000
0x48800000
0x88800000

It's actually super important that if I write to 0x48800000 (which is commonly used to bypass CPU caches for interop with GPU or etc.) that a later read from 0x08800000 will read exactly the same value.

Normally, this is made to work by passing the same `fd, offset` parameters for both, but different addresses so we can setup this mapping.  If it allocates where ever it wants (which is what null means here), then these later mirrors won't work, and I'd expect some things to not work right in games.

-[Unknown]

Reply:
Thats true, I thought about that and forgot to go back and look at this. On runtime without this I'd get an invalid argument error, did this temporarily to get it to at least run, and forgot to look at it. I'll look at this

Reply:
On platforms where we can't set up mirror mappings (or don't have enough address space) we apply a different workaround by simply &-ing every address used for a memory access with 0x3FFFFFFF. See MASKED_PSP_MEMORY

Reply:
So basically, the way psp works is it creates a 4GB file, which is the emulated psp's memory, and maps that into memory, correct? And enabling `MASKED_PSP_MEMORY` keeps the memory in memory? Did I understand that correctly?

Reply:
As I mentioned, we want multiple pointers to point to the same data.  The good news is: most operating systems already have a way to do this.  It's faster to just let the operating system handle it, instead of doing it manually.

But it's not like operating systems have a function to just map memory to two addresses.  That'd be too easy and convenient.  No, what they have is a tool that I believe was originally meant for inter-process communication: shared file memory mapping.  It just maps shared memory, but we can map the "file" to multiple places and it all points to the same RAM.  This is often supported, but only works if we can get 4 GB of address space reserved all in a row.

On 32-bit operating systems, that's not possible.  We still use the mapping for VRAM there (which has multiple mirrors inside 0x3FFFFFFF of VRAM), but we allocate less address space, and the mirrors I mentioned above aren't "needed" because we mask pointers instead.

So if emscripten doesn't support shared memory mapping (maybe because they almost explicitly don't want inter-process communication?), your best bet is to define MASKED_PSP_MEMORY (see where it's defined for 32-bit) on emscripten and hope the VRAM mirrors don't cause problems.  Most of the time they'd only be used for depth reads and writes from the CPU.

-[Unknown]